Implement change tracking support for JSON mapped properties

lauxjpn opened the pull request:
While the bug was most apparent with POCO array elements and POCO properties in general, this fix applies to all JSON mapped properties (string, DOM and POCO).

Relates to #1102, #1157 and #1158
Fixes #1140

AndriySvyryd commented on src/EFCore.MySql/Storage/ValueComparison/Internal/JsonValueComparer.cs:
```diff
@@ -0,0 +1,16 @@
+            : base(
+                (left, right) => object.Equals(valueConverter.ConvertToProvider(left), valueConverter.ConvertToProvider(right)),
+                value => value.GetHashCode(),
+                value => (T)valueConverter.ConvertFromProvider(valueConverter.ConvertToProvider(value)))
```
The snapshot lambda is invoked for every property when the entity is tracked or is saved. Making it faster would result in noticeable perf improvements.

lauxjpn replied:
It could be improved for _some_ scenarios as far as I can see:
* Newtonsoft.Json if `T` is a DOM class: Use [JTokenEqualityComparer](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Linq_JTokenEqualityComparer.htm)
* Any Json-Library if `T` is `string`: Just compare the strings as is.

This would however do nothing for the most common case, which should be to use POCOs.

AndriySvyryd replied:
You could check for `ICloneable` implementation, `JToken` has `DeepClone` and I believe JsonDocument is immutable, so same value can be returned

lauxjpn replied:
> `JToken` has `DeepClone` and I believe `JsonDocument` is immutable

Good points!

> You could check for `ICloneable` implementation

Theoretically true, though the `ICloneable` contract is traditionally not explicit in _how_ the object is being cloned. From the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.icloneable?view=netcore-3.1#notes-to-implementers):

> Because callers of `Clone()` cannot depend on the method performing a predictable cloning operation, we recommend that `ICloneable` not be implemented in public APIs.

AndriySvyryd replied:
You could also look for `DeepClone`, as long as this is documented it will allow to implement it efficiently for POCOs

lauxjpn replied:
> You could also look for `DeepClone()`, as long as this is documented it will allow to implement it efficiently for POCOs

Could be a better approach. As for `ICloneable`, we might make this opt-in/opt-out.

roji replied:
While JsonDocument is immutable and so there's no need to actually snapshot or do anything beyond reference comparison, JsonElement is a struct, and so that's not really possible... We should consider not allowing to map JsonElement.

lauxjpn replied:
I agree that it makes little sense to snapshot `JsonElement` with the default struct member by member copy in the way `JsonElement` has been implemented. If the `JsonElement._parent` field had a publicly available property, this wouldn't be a problem, since the parent would still be immutable. But for some reason (no idea why) it has no property.

roji replied:
You're right (https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1140#issuecomment-691620821)

AndriySvyryd commented on src/EFCore.MySql/Storage/ValueComparison/Internal/JsonValueComparer.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        public JsonValueComparer(ValueConverter valueConverter)
+            : base(
+                (left, right) => object.Equals(valueConverter.ConvertToProvider(left), valueConverter.ConvertToProvider(right)),
```
Equals lambda is also on the hot path.

lauxjpn replied:
The only general way I could think about speeding this up was to traverse the object graphs ourselves in parallel, so we can compare them node by node and possibly break early, if they diverge. This would however be significantly more complex then just letting the JSON library do the work and comparing the result.

For Newtonsoft.Json, if `T` is a DOM class, then [JTokenEqualityComparer](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Linq_JTokenEqualityComparer.htm) could be used.

AndriySvyryd replied:
You could detect whether it implements `IEquatable<T>` or overrides `Equals` and use that.

lauxjpn replied:
> You could detect whether `T` implements `IEquatable<T>` or overrides `Equals` and use that.

Good points!

lauxjpn commented:
However, the moment we at least don't process JSON mappings that are to a `T` of `string`, users would have at least an implicit way to improve performance by using `string` instead of a POCO.

lauxjpn commented:
I added the optimized change tracking support for `System.Text.Json` and `Newtonsoft.Json`.

The `UseMicrosoftJson` and `UseNewtonsoftJson` extension methods now contain an optional `MySqlCommonJsonChangeTrackingOptions` parameter, that can be used to specify the global default change tracking options used for all JSON column mapped properties. This should help immensely with the discoverability of the change tracking options. The default value is `RootPropertyOnly`.

Currently, we only have one value comparer for each JSON support library, the handles the expression construction for everything JSON. We can split it up later if necessary (like we have for the JSON value converters). This should not result in backwards compatibility issue, because we marked the value comparers (as well as the value converters) as _internal_.

The value comparers (and converters) are mapped automatically with the global default change tracking options. If a user wants to change the options for specific model properties, we introduced an `UseJsonChangeTrackingOptions` property builder extension method per JSON support library.

So unless users don't explicitly opt into internal behavior, the internals have been abstracted away. For EF Core 6 and its possible JSON out-of-the-box support, this gives us a lot of flexibility. E.g. we could alter the internal in accordance to new requirements and then make them publicly available, while keeping our current interface.

The publicly available `MySqlCommonJsonChangeTrackingOptions` are common combinations of the internal `MySqlJsonChangeTrackingOptions` flags, that finely control how comparison, snapshotting and hashing is handled.

sukhdev22 commented:
Surprisingly, I don't see a single **Sample application** which showcases how to map JSON columns to POCO .NET properties. How to insert, update and query Out-Of-The-Box using this library. Can anyone point to any such resource ?

lauxjpn commented:
@sukhdev22 The most comprehensive sample code we currently have is probably the one from [#14 (comment)](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/14#issuecomment-725257107).

It demonstrates all three types of JSON support: POCOs, the DOM and strings.

mguinness commented:
Yes, both documentation and sample code could be improved and are highlighted in https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/522 and https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/935.

There is also https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1179 which asks for help in that effort, so if you could contribute it would be welcomed.
